pkg/controllers/nodeclass/capacityreservation.go

```diff
@@ -0,0 +1,133 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclass
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/capacityreservation"
+)
+
+const capacityReservationPollPeriod = time.Minute
+
+type CapacityReservation struct {
+	provider capacityreservation.Provider
+	clk      clock.Clock
+	cm       *pretty.ChangeMonitor
+}
+
+func NewCapacityReservationReconciler(clk clock.Clock, provider capacityreservation.Provider) *CapacityReservation {
+	return &CapacityReservation{
+		provider: provider,
+		clk:      clk,
+		cm:       pretty.NewChangeMonitor(),
+	}
+}
+
+func (c *CapacityReservation) Reconcile(ctx context.Context, nc *v1.EC2NodeClass) (reconcile.Result, error) {
+	reservations, err := c.provider.List(ctx, nc.Spec.CapacityReservationSelectorTerms...)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting capacity reservations, %w", err)
+	}
+	if len(reservations) == 0 {
+		nc.StatusConditions().SetTrue(v1.ConditionTypeCapacityReservationsReady)
+		return reconcile.Result{RequeueAfter: capacityReservationPollPeriod}, nil
```

Comment:
nit: I could go either way about whether to inline this or not but I know that it's all up to personal preference 🤷🏼‍♂️

Reply:
It's used in a couple places, so I'd rather leave it as is.